### ISSUE TITLE
YAML implicitly converts fields to other datatypes

### ISSUE THREAD

**Issue by Fokko** [fishtown-analytics/dbt#2175]
### Describe the bug

With Google Analytics, some tables are numeric, these table names are automagically converted to bool/int/Decimal.

```
MacBook-Pro-van-Fokko:dbt-source fokkodriesprong$ python3
Python 3.6.6 (v3.6.6:4cf1f54eb7, Jun 26 2018, 19:50:54)
[GCC 4.2.1 Compatible Apple LLVM 6.0 (clang-600.0.57)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import yaml
>>>
>>> yaml.load("name: 19.25")
{'name': 19.25}
>>> data = yaml.load("name: 19.25")
>>> type(data['name'])
<class 'float'>

>>> data = yaml.load("name: 1925")
>>> type(data['name'])
<class 'int'>

>>> data = yaml.load("name: 'no'")
>>> type(data['name'])
<class 'str'>

>>> data = yaml.load("name: no")
>>> type(data['name'])
<class 'bool'>
```

### Steps To Reproduce

Have a source with a Decimal name.
```
  - name: 19.25
    schema: "{{ var('test_run_schema') }}"
    quoting:
      identifier: True
    tables:
      - name: test_table
        identifier: other_source_table
```

### Screenshots and log output

```
2020-03-02 09:10:33,888249 (MainThread): 'decimal.Decimal' object has no attribute 'lower'
2020-03-02 09:10:33,892358 (MainThread): Traceback (most recent call last):
  File "/usr/local/Cellar/dbt/0.15.2_1/libexec/lib/python3.7/site-packages/dbt/main.py", line 80, in main
    results, succeeded = handle_and_check(args)
  File "/usr/local/Cellar/dbt/0.15.2_1/libexec/lib/python3.7/site-packages/dbt/main.py", line 158, in handle_and_check
    task, res = run_from_args(parsed)
  File "/usr/local/Cellar/dbt/0.15.2_1/libexec/lib/python3.7/site-packages/dbt/main.py", line 210, in run_from_args
    results = task.run()
  File "/usr/local/Cellar/dbt/0.15.2_1/libexec/lib/python3.7/site-packages/dbt/task/generate.py", line 201, in run
    catalog_table = adapter.get_catalog(self.manifest)
  File "/usr/local/Cellar/dbt/0.15.2_1/libexec/lib/python3.7/site-packages/dbt/adapters/base/impl.py", line 1016, in get_catalog
    results = self._catalog_filter_table(table, manifest)
  File "/usr/local/Cellar/dbt/0.15.2_1/libexec/lib/python3.7/site-packages/dbt/adapters/bigquery/impl.py", line 493, in _catalog_filter_table
    return super()._catalog_filter_table(table, manifest)
  File "/usr/local/Cellar/dbt/0.15.2_1/libexec/lib/python3.7/site-packages/dbt/adapters/base/impl.py", line 998, in _catalog_filter_table
    return table.where(_catalog_filter_schemas(manifest))
  File "/usr/local/Cellar/dbt/0.15.2_1/libexec/lib/python3.7/site-packages/agate/table/where.py", line 25, in where
    if test(row):
  File "/usr/local/Cellar/dbt/0.15.2_1/libexec/lib/python3.7/site-packages/dbt/adapters/base/impl.py", line 82, in test
    return (table_database.lower(), table_schema.lower()) in schemas
AttributeError: 'decimal.Decimal' object has no attribute 'lower'
```

### System information
**Which database are you using dbt with?**
- [ ] postgres
- [ ] redshift
- [x] bigquery
- [ ] snowflake
- [ ] other (specify: ____________)

**The output of `dbt --version`:**
```
2020-03-02 09:10:29,518713 (MainThread): On generate_catalog: /* {"app": "dbt", "dbt_version": "0.15.2", "profile_name": "XXXX", "target_name": "prod", "connection_name": "generate_catalog"} */
```

**The operating system you're using:**

OSX

**The output of `python --version`:**

```
MacBook-Pro-van-Fokko:dbt-source fokkodriesprong$ python3 --version
Python 3.6.6
```

**Comment by drewbanin** [comment]
hey @Fokko - thanks for opening this issue! I see how a numeric table name could cause issues with the code that dbt uses to operate on relations:
```
return (table_database.lower(), table_schema.lower()) in schemas
```

Do you have control over the schema.yml files that declare these sources? My recommendation would be to supply the table name as a string, rather than as a decimal. You're not seeing implicit conversion in the example shared above:
```
  - name: 19.25
    schema: "{{ var('test_run_schema') }}"
    quoting:
      identifier: True
    tables:
      - name: test_table
        identifier: other_source_table
```

In this case, you're just supplying a number where you should be supplying a string. Can you do something like:
```
  - name: "19.25" # <----- note the quotes here!
    schema: "{{ var('test_run_schema') }}"
    quoting:
      identifier: True
    tables:
      - name: test_table
        identifier: other_source_table
```

If I understand everything correctly, this should solve you're problem!

**Comment by Fokko** [comment]
Thanks for pointing this out Drew. Actually I'm helping someone on the devlist: https://getdbt.slack.com/archives/C2JRRQDTL/p1582909962173000

The weird thing that I'm not able to reproduce the issue. It seems that there is already code in place that enforces this, hence the failing CI.

**Comment by dalebradman** [comment]
That person is me 👋 .

I have stripped the project right back to have:

- The out-the-box `dbt_project.yml` template that links to my `~/.dbt/profiles.yml` file
- An empty `schema.yml` file
- A single model that uses no references like so:

```
SELECT * FROM `frontend-data.200111222.sessions_*`
```

But still running `dbt docs generate` fails with the same:

```
Encountered an error:
'decimal.Decimal' object has no attribute 'lower'
```

**Comment by dalebradman** [comment]
I’ve installed 0.16.0-rc2. When I run `dbt docs generate` I still get the same error however it also is able to write down the `catalog.json` file and therefore running dbt docs serve works:
```
✗ dbt docs generate
Running with dbt=0.16.0-rc2
Found 5 models, 32 tests, 0 snapshots, 0 analyses, 268 macros, 0 operations, 0 seed files, 1 source
11:56:13 | Concurrency: 1 threads (target='prod')
11:56:13 |
11:56:14 | Done.
11:56:14 | Building catalog
Encountered an error while generating catalog: 'decimal.Decimal' object has no attribute 'lower'
dbt encountered 1 failure while writing the catalog
11:56:26 | Catalog written to /Users/DbtProjects/data-pipelines/frontend/target/catalog.json
```

**Comment by drewbanin** [comment]
ooh! @dalebradman - I was really stumped here, but I think i know what this is :D 

Check out the fix for this here: https://github.com/fishtown-analytics/dbt/pull/2208

This fix is shipping in 0.16.0, likely by the end of this week :)

**Comment by drewbanin** [comment]
closed via https://github.com/fishtown-analytics/dbt/pull/2208
